src/main/kotlin/apoc/nlp/aws/AWSProcedures.kt

```diff
@@ -0,0 +1,123 @@
+
+    @Context
+    @JvmField
+    var tx: Transaction? = null
+
+    @Procedure(value = "apoc.nlp.aws.entities.stream", mode = Mode.READ)
+    @Description("Returns a stream of entities for provided text")
+    fun entitiesStream(@Name("source") source: Any,
+                 @Name(value = "config", defaultValue = "{}") config: Map<String, Any>) : Stream<NodeWithMapResult> {
+        verifySource(source)
+        val nodeProperty = getNodeProperty(config)
+        verifyNodeProperty(source, nodeProperty)
+        verifyKey(config, "key")
+        verifyKey(config, "secret")
+
+        val client: AWSClient = RealAWSClient(config, log!!)
+        val detectEntitiesResult = client.entities(source)
+
+        return convert(source).mapIndexed { index, node -> transformResults(index, node, detectEntitiesResult!!) }.stream()
+    }
+
+    @Procedure(value = "apoc.nlp.aws.entities.graph", mode = Mode.WRITE)
+    @Description("Creates a (virtual) entity graph for provided text")
+    fun entitiesGraph(@Name("source") source: Any,
+                      @Name(value = "config", defaultValue = "{}") config: Map<String, Any>) : Stream<VirtualGraph> {
+        verifySource(source)
+        val nodeProperty = getNodeProperty(config)
+        verifyNodeProperty(source, nodeProperty)
+        verifyKey(config, "key")
+        verifyKey(config, "secret")
+
+        val client = RealAWSClient(config, log!!)
+        val detectEntitiesResult = client.entities(source)
+
+        val storeGraph: Boolean = config.getOrDefault("write", false) as Boolean
+
+        val relationshipType = NLPHelperFunctions.entityRelationshipType(config)
+
+        val virtualNLPGraph = AWSVirtualNLPGraph(detectEntitiesResult!!, convert(source), relationshipType, ENTITY_MAPPING)
+        return if (storeGraph) {
+            Stream.of(virtualNLPGraph.createAndStore(tx))
+        } else {
+            Stream.of(virtualNLPGraph.create())
+        }
+    }
+
+    companion object {
+        fun transformResults(index: Int, node: Node, res: BatchDetectEntitiesResult): NodeWithMapResult {
+            val result = res.resultList.find { result -> result.index == index }
+            return if (result != null) {
+                NodeWithMapResult.withResult(node, JsonUtil.OBJECT_MAPPER!!.convertValue(result, Map::class.java) as Map<String, Any?>)
+            } else {
+                val err = res.errorList.find { error -> error.index == index }
+                NodeWithMapResult.withError(node, mapOf("code" to err?.errorCode, "message" to err?.errorMessage))
+            }
+        }
+
+        private fun convert(source: Any): List<Node> {
+            return when (source) {
+                is Node -> listOf(source)
+                is List<*> -> source.map { item -> item as Node }
+                else -> throw java.lang.IllegalArgumentException("`source` must be a node or list of nodes, but was: `${source}`")
+            }
+        }
+
+        private fun verifySource(source: Any) {
```

Comment:
can we move this and the other validation methods to Util class and reuse them in AWS and GCP procedures?